### PR TITLE
fix(generate): correct component path when module is generated in subfolder, and parent folder is not a module too

### LINKED ISSUE #3255: Creating a module deeper than src/app throws Invalid path
> Please provide us with the following information:
> ---------------------------------------------------------------

   node: 6.7.0
   os: darwin x64

### Repro steps.
- Create new project with ng new my-app
- Create a new folder inside src/app (e.g. src/app/modules)
- Go to new folder (cd src/app/modules)
- Create new module (ng g module core)

### The log given by the failure.
Invalid path: "<my-project-path>/src/app/modules/modules"" is not a valid path.

### Mention any other details that might be useful.
To summarise this issue. While creating a new module the cli tries to go into the same folder again, if the module is created deeper than /src/app. Creating an module inside /src/app works as expected.

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

### PR DESCRIPTION
**Pull request by Meligy**
fixes #3255

Another attempt at using the correct component name / path when generating modules, that takes care of the special case in #3255.

## Existing behaviour that this PR fixes

if you `ng g module parent --routing`, you can `ng g module parent/child  --routing`.

But if `src/app/parent` is just a normal folder, the component generation for `ng g module parent/child` fails.

## To test the PR:

- Create a new project linked to the CLI from the PR branch
- Inside the project, run `ng g module normal-case --routing` to ensure basic scenario is not broken
- Then also run `ng g module parent` and `ng g module parent/child --routing` to ensure this scenario is not broken too
- Then run `mkdir src/app/parent-folder`, and then `ng g module parent-folder/child-module --routing` to test that this PR does what it promises to do

Or alternatively look at the provided tests in the PR and existing tests, and ensure they cover all above scenarios, and are passing in CI.

## Note

This is maybe the 3rd or so attempt at getting this path to work in all scenarios. We are protected by tests though, and we keep adding more tests the more scenarios we get. Let's see if this becomes the last change of this path/name.

### CODE REVIEW

**Comment by hansl** [comment]
Could you also add an e2e test? Acceptance tests are good but likely to be removed after 1.0 but e2e will remain.

**Comment by Meligy** [comment]
@hansl  Added e2e test

**Comment by Meligy** [comment]
e2e test is passing now.

**Comment by hansl** [comment]
LGTM. Need @Brocco's approval.

File: packages/angular-cli/blueprints/module/index.js [hunk 1]
```diff
@@ -58,13 +58,20 @@ module.exports = {
...
-      options.entity.name = path.relative(this.dynamicPath.appRoot, componentPath);
+
+      // Component folder needs to be `/{moduleName}/{ComponentName}`
+      // Note that we are using `flat`, so no extra dir will be created
+      // We need the leading `/` so the component path resolution work for both cases below:
+      // 1. If module name has no path (no `/`), that's going to be `/mod-name/mod-name`
+      //      as `this.dynamicPath.dir` will be the same as `this.dynamicPath.appRoot`
+      // 2. If it does have `/` (like `parent/mod-name`), it'll be `/parent/mod-name/mod-name`
+      //      as `this.dynamicPath.dir` minus `this.dynamicPath.appRoot` will be `/parent`
+      var dmouleDir =
```
Comment (Brocco) [inline packages/angular-cli/blueprints/module/index.js]: typo? rename please

`dmouleDir` -> `moduleDir`

Comment (Meligy) [inline packages/angular-cli/blueprints/module/index.js]: Done.

**Comment by Meligy** [comment]
Can we mark this for merge now?

Cheers,

**Pull request merged**